GVFS/GVFS.Platform.Linux/LinuxFileSystemVirtualizer.cs

```diff
@@ -0,0 +1,719 @@
+            {
+                string symLinkTarget;
+                if (this.TryGetSymLinkTarget(shaContentId, out symLinkTarget))
+                {
+                    Result result = this.virtualizationInstance.ReplacePlaceholderFileWithSymLink(
+                        relativePath,
+                        symLinkTarget,
+                        (UpdateType)updateFlags,
+                        out failureCause);
+
+                    this.FileSystemCallbacks.OnFileSymLinkCreated(relativePath);
+
+                    failureReason = (UpdateFailureReason)failureCause;
+                    return new FileSystemResult(ResultToFSResult(result), unchecked((int)result));
+                }
+
+                EventMetadata metadata = this.CreateEventMetadata(relativePath);
+                metadata.Add(nameof(shaContentId), shaContentId);
+                this.Context.Tracer.RelatedError(metadata, $"{nameof(this.UpdatePlaceholderIfNeeded)}: Failed to read contents of symlink object");
+                failureReason = UpdateFailureReason.NoFailure;
+                return new FileSystemResult(FSResult.IOError, 0);
+            }
+            else
+            {
+                EventMetadata metadata = this.CreateEventMetadata(relativePath);
+                metadata.Add(nameof(fileType), fileType);
+                metadata.Add(nameof(fileMode), fileMode);
+                this.Context.Tracer.RelatedError(metadata, $"{nameof(this.UpdatePlaceholderIfNeeded)}: Unsupported fileType");
+                failureReason = UpdateFailureReason.NoFailure;
+                return new FileSystemResult(FSResult.IOError, 0);
+            }
+        }
+
+        protected override bool TryStart(out string error)
+        {
+            error = string.Empty;
+
+            // Callbacks
+            this.virtualizationInstance.OnEnumerateDirectory = this.OnEnumerateDirectory;
+            this.virtualizationInstance.OnGetFileStream = this.OnGetFileStream;
+            this.virtualizationInstance.OnLogError = this.OnLogError;
+            this.virtualizationInstance.OnFileModified = this.OnFileModified;
+            this.virtualizationInstance.OnPreDelete = this.OnPreDelete;
+            this.virtualizationInstance.OnNewFileCreated = this.OnNewFileCreated;
+            this.virtualizationInstance.OnFileRenamed = this.OnFileRenamed;
+            this.virtualizationInstance.OnHardLinkCreated = this.OnHardLinkCreated;
+            this.virtualizationInstance.OnFilePreConvertToFull = this.NotifyFilePreConvertToFull;
+
+            uint threadCount = (uint)Environment.ProcessorCount * 2;
+
+            Result result = this.virtualizationInstance.StartVirtualizationInstance(
+                this.Context.Enlistment.WorkingDirectoryBackingRoot,
+                this.Context.Enlistment.WorkingDirectoryRoot,
+                threadCount,
+                this.IsUninitializedMount(this.Context.Enlistment.WorkingDirectoryBackingRoot));
```

Comment:
Full disclosure: I'm still very new to the Linux architecture.

If `IsUninitializedMount` is false, should `StartVirtualizationInstance` be called at all?

Reply:
Good question -- in brief, yes, because the issue here is that we need the provider to inform us whether the top-level directory of a mount (or a remount of an unmounted working directory) should have its empty placeholder flag set or cleared.  The underlying "backing root" (e.g., `.vfsforgit/lower`) is created by the clone step (`gvfs clone`), but it has no placeholder flag set.

A user might mount (`gvfs mount`), visit the working directory and cause a hydration, and unmount(`gvfs unmount`), and then remount later, and we do want to start the virtualization again on the second `gvfs mount`.

Your question does make me wonder if we should choose a more descriptive name for the `IsUninitializedMount()` check ... @kivikakk, any thoughts here?  (Naming is never my strong suit! :-)

Reply:
I don't have any strong opinions on the name — `MountNotYetHydrated`? Something that conveys a little more specifically what the test is for would be ideal.

Reply:
I'm going to go with `IsDirectoryEmpty()` and `IsWorkingDirectoryEmpty()` for `MirrorProvider` and `GVFS` respectively -- I think that captures (a) the difference between them, (b) the fact that we're testing whether the directory is "empty" (from the provider's standpoint) and needs to be flagged for future hydration, and (c) that we're really only testing a directory's contents.  The fact that it's a mount point is incidental to the tests.  I hope that works for everyone, and if not, we can change again!